com.teamscale.polarion.plugin/src/com/teamscale/polarion/plugin/WorkItemUpdatesServlet.java

```
@@ -0,0 +1,769 @@
+        if (Integer.valueOf(workItemHistory.get(0).getRevision()) <= Integer.valueOf(endRevision))
+          workItemForJson =
+              Utils.castWorkItem(workItemHistory.get(0), includeCustomFields, includeLinkRoles);
+      } else if (workItemHistory.size() > 1) {
+        /**
+         * From Polarion JavaDoc: "The history list is sorted from the oldest (first) to the newest
+         * (last)."
+         * https://almdemo.polarion.com/polarion/sdk/doc/javadoc/com/polarion/platform/persistence/IDataService.html#getObjectHistory(T)
+         * Then, we get the last one from the history as the current revision
+         */
+        int lastUpdateIndex = searchIndexWorkItemHistory(workItemHistory);
+        IDiffManager diffManager = dataService.getDiffManager();
+        Collection<WorkItemChange> workItemChanges = new ArrayList<WorkItemChange>();
+        int endIndex =
+            collectWorkItemChanges(
+                workItemChanges, workItem.getId(), workItemHistory, diffManager, lastUpdateIndex);
+        // Using the endIndex to return the workItem as in the endRevision # (not necessarily the
+        // latest version of the item)
+        if (endIndex >= 0) {
+          workItemForJson =
+              Utils.castWorkItem(
+                  workItemHistory.get(endIndex), includeCustomFields, includeLinkRoles);
+          workItemForJson.setWorkItemChanges(workItemChanges);
+        }
+      } else {
+        /**
+         * No history. Empty list. From Polarion JavaDoc: "An empty list is returned if the object
+         * does not support history retrieval."
+         * "https://almdemo.polarion.com/polarion/sdk/doc/javadoc/com/polarion/platform/persistence/IDataService.html#getObjectHistory(T)"
+         */
+        // Since we're grabbing WIs (and they support history retrieval), as far as we can tell from
+        // Polarion docs, this else will never execute.
```

If this should never happen please throw an exception here

---

Yeah, if that ever happens it's a Polarion server failure. So, I raised an exception and on the outermost method I catch, log, and return a 500 error to the client. The rationale is: if we cannot return the history of an item (which is that else), it's better to not fulfill the request and return a server error rather than skip that workr item and return a state that does not necessarily reflect the work item history and potentially lead to inconsistencies on the client side.

Done.